### ISSUE TITLE
GAT defaults to where Self: 'a code breaking report

### ISSUE THREAD

**Issue by chubei-oppen** [rust-lang/rust#92033]
As suggested in #87479, I'm reporting a case where the required bound breaks code.

I was writing a 3D renderer and modeling the `surface` and `swapchain` concepts using traits.

A `Surface` is something you can render to, it should produce a iterator of references to `Texture`s:

```rust
struct Texture;

trait Surface {
    type TextureIter<'a>: Iterator<Item = &'a Texture>;

    fn get_texture(&self) -> Self::TextureIter<'_>;
}
```

A `Swapchain` produces a reference to a `Surface`:

```rust
trait Swapchain {
    type Surface<'a>: Surface;

    fn get_surface(&self) -> Self::Surface<'_>;
}
```

Now a `Texture` itself can be used as a swapchain, producing reference to itself as the `Surface`.

```rust
impl<'s> Surface for &'s Texture {
    type TextureIter<'a> = std::option::IntoIter<&'a Texture>;

    fn get_texture(&self) -> Self::TextureIter<'_> {
        let option: Option<&Texture> = Some(self);
        option.into_iter()
    }
}

impl Swapchain for Texture {
    type Surface<'a> = &'a Texture;

    fn get_surface(&self) -> Self::Surface<'_> {
        self
    }
}
```

Above code compiles under `rustc 1.58.0-nightly (0d1754e8b 2021-11-05)`.

Now it doesn't with added `where Self: 'a` bound.

This is a simplified example as `Surface` and `Swapchain` have other required methods.

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (efec54529 2021-12-04
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
error[E0477]: the type `&'s Texture` does not fulfill the required lifetime
  --> src/main.rs:18:5
   |
18 |     type TextureIter<'a> = std::option::IntoIter<&'a Texture>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: type must outlive the lifetime `'a` as defined here
  --> src/main.rs:18:22
   |
18 |     type TextureIter<'a> = std::option::IntoIter<&'a Texture>;
   |                      ^^

For more information about this error, try `rustc --explain E0477`.
error: could not compile `rsplayground` due to previous error
```

</p>
</details>

**Comment by jackh726** [comment]
I think this can be solved by just added the `where Self: 'a` bounds to the impl too...

**Comment by jackh726** [comment]
That being said, I feel like this *should* be fine, so there might be something more subtle going on.

**Comment by chubei-oppen** [comment]
> I think this can be solved by just added the `where Self: 'a` bounds to the impl too...

How to add `where Self: 'a` to the impl? Is it after `impl<'s> Surface for &'s Texture` or `type TextureIter<'a> = std::option::IntoIter<&'a Texture>`?

**Comment by jackh726** [comment]
This works (with the where clause on the impl): https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=cc1325c7ca369536e63c9e436d6eacf0

Were you trying to put the where clause on the impl at the end of the line? Like `type TextureIter<'a> = std::option::IntoIter<&'a Texture> where Self: 'a`?

**Comment by chubei-oppen** [comment]
Yes... Should have thought of the correct place. It works in real code too. Closing the issue. Thanks a lot!

**Comment by jackh726** [comment]
I'm actually going to leave this open for two diagnostics improvements:
1) Proper suggestion to move the where clause if it's placed in the wrong place on the GAT (#90076 changes the location, but I'm about to open a PR that doesn't change the location but parses/errors the invalid location)
2) Maybe try to suggest `Self: 'a` on the impl GAT in some way. Either through suggest copying the trait bounds, or recognizing that `&'s Texture` is the self type.

**Comment by jackh726** [comment]
GATs issue triage: not blocking. At this point, this is just a diagnostics improvement. 1 from above is fixed; just want to do 2.
